### DISCUSSION TITLE
Construct temporary table from existing ORM model

### DISCUSSION THREAD

**Discussion by Diego-Hernandez-Moodys** [sqlalchemy/sqlalchemy#9400] [Usage Questions] [answered]
I'd like to create a staging table that has the exact same schema as the final table for quick insertions and check existence, filter out records, etc.

Assume an instance_session and instance_table are available. If I can retrieve the engine used to create the session, that would be helpful.

The CORE way wasn't so simple, but it got the job done.
```python

# Create CORE table
Meta = sqlalchemy.MetaData(bind = None) # could try to extract engine from session
list_instance_columns = [column.copy() for column in instance_table.__table__.columns.values()] # sets table value to None
Staging = sqlalchemy.Table("staging", Meta, *list_instance_columns, prefixes=['TEMPORARY'], schema=instance_table.__bind_key__)

statement_create = f"CREATE TEMPORARY TABLE {instance_table.__bind_key__}.staging SELECT * FROM {instance_table.__bind_key__}.{instance_table.__tablename__} LIMIT 0"

instance_session.execute(statement_create)
statement_insertion = Staging.insert().values(**dict_data)
instance_session.execute(statement_insertion)
```
This works, but I have to insert millions of rows, so calling the insert().values(**dict_params) is probably too slow as I'd like to collect the row objects then batch insert.

My question is how do I create an equivalent ORM constructor for this temp table? Where could I potentially pass the columns? Or, could I inherit from the instance_table without modifying its ORM model? Ideally, I'd like to ignore the class altogether since and just use the constructor interface since it's just a temp table, but this workaround works too.
```python
class Staging(Base):
    __tablename__ = "staging"
    __bind_key__ "database-name"
    __table_args__ = {'prefixes': ['TEMPORARY']}
 
list_rows = []
for dict_data in list_dicts_data:
    list_rows.append(Staging(**dict_data))
# pass them onto some other function
```

**Comment by CaselIT** [comment, marked as answer]
hi,

You could connect the table to the class, setting the `__table__` instead of `__tablename__`. See https://docs.sqlalchemy.org/en/20/orm/declarative_tables.html#declarative-with-imperative-table-a-k-a-hybrid-declarative

**Comment by zzzeek** [comment]
OK don't use column.copy(), which is deprecated as it was not intended to be public.  Use [Table.to_metadata()](https://docs.sqlalchemy.org/en/20/core/metadata.html#sqlalchemy.schema.Table.to_metadata), which lets you change both the name and the schema of the table to something new.  You can add the "TEMPORARY" thing after the fact to '_prefixes' and we can support adding this argument to the `to_metadata()` call as a new use case.

then as @CaselIT  mentions, with your new `Table` you can associate it with a new declarative class via ``__table__``.

**Reply by Diego-Hernandez-Moodys** [reply]
So if I read you correctly,
```python
Base = sqlalchemy.ext.declarative.declarative_base()
TableStaging = instance_table.__table__.to_metadata(Base.metadata, schema=database, name = "staging")
class Staging(Base):
    __tablename__ = "staging"
    __table__ = TableStaging
    __table_args__ = {'prefixes': ['TEMPORARY']}
```
This works well, thank you.

Now, what's the best way to reuse the Base from the existing table? I feel like having to create an empty instances of Base for one-time-use is anti-pattern.

Is there a better way to create this new staging table from the existing session? This is what I currently use, but I'd like to stick to using sqlachemy and create the table from the ORM class. I don't know if I have access to the engine directly unless I use something like `engine = session.get_bind()`
```python
statement = f"CREATE TEMPORARY TABLE {database}.staging SELECT * FROM {instance_table.__tablename__} LIMIT 0"
instance_session.execute(statement)
```

**Reply by Diego-Hernandez-Moodys** [reply]
I also went into the debugging session and noticed the new table Staging actually points to the original table
```
  Base.metadata.create_all(instance_session.get_bind(),checkfirst=True)
  d = {"primary_key": 5e6, "column": "FAKE"}
  instance_session.add(Staging(**d))
  instance_session.commit()
```

**Reply by zzzeek** [reply]
you can use the Base you have, as long as your table name and/or schema is different, just note there's no "remove" operation so you'd want to hold onto that same Staging class for as long as you need it.    Otherwise if you need Staging to go away, then you would make either a new `registry` and/or `Base` to map the class temporarily.

to create the table in the current transaction use `Session.connection()`:

```py
Staging.__table__.create(instance_session.connection())
```

> I also went into the debugging session and noticed the new table Staging actually points to the original table

the meaning of this sentence is ambiguous.

**Reply by Diego-Hernandez-Moodys** [reply]
Despite the name attributes being called staging, from the debug console
```
Staging.__table__.columns[0]
>>> Column('id', INTEGER(display_width=11), table=<staging>, primary_key=True, nullable=False)
```
The commit log says the new record was committed to the original table.

**Reply by zzzeek** [reply]
Make sure `Staging.__table__.name` and `Staging.__table__.schema` are what's expected.  beyond that, data going into the wrong place would involve however these classes are being used.    Also, I think "bind_key" is a Flask thing?   Not sure what kinds of things flask might be doing.

**Reply by zzzeek** [reply]
note also this wont work:

```py
Base = sqlalchemy.ext.declarative.declarative_base()
TableStaging = instance_table.__table__.to_metadata(Base.metadata, schema=database, name = "staging")
class Staging(Base):
    __tablename__ = "staging"
    __table__ = TableStaging
    __table_args__ = {'prefixes': ['TEMPORARY']} 
```

those tableargs are ignored.   as mentioned previously, we don't have support for custom prefixes in to_metadata so you need to use private API for now:


```py
Base = sqlalchemy.ext.declarative.declarative_base()
TableStaging = instance_table.__table__.to_metadata(Base.metadata, schema=database, name = "staging")
TableStaging._prefixes = ["TEMPORARY"]
class Staging(Base):
    __table__ = TableStaging
```

Here is a runnable demo

```py
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import Session
Base = declarative_base()


class A(Base):
    __tablename__ = 'a'

    id = Column(Integer, primary_key=True)
    data = Column(String)

e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

# if using the same Base, hold onto TableStaging and Staging
# as long as Base exists
TableStaging = A.__table__.to_metadata(Base.metadata, name = "staging")
TableStaging._prefixes = ['TEMPORARY']

class Staging(Base):
    __table__ = TableStaging


with Session(e) as sess:
    sess.add(A(data='d1'))
    sess.flush()

    TableStaging.create(sess.connection())

    sess.add(Staging(data='staging data'))

    sess.commit()


```

**Reply by Diego-Hernandez-Moodys** [reply]
The __bind_key__ is just a database name. I like how Flask uses it, but it has no affect on the final product. I'll see what setting the schema to None does.

**Reply by Diego-Hernandez-Moodys** [reply]
> we don't have support for custom prefixes in to_metadata
That's fine.

It worked. Thanks so much. This goes a long way to document a means to create a temp table via SQLAlchemy.

Is there an obvious way to see which database and table name the ORM class is associated with? As in, any way to verify the __tablename__ attribute is the same as what's in the actual database?

**Reply by zzzeek** [reply]
sure, `class.__table__`, that's the `Table` object, it has a `name`, `schema`, that's it
